californium-core/src/main/java/org/eclipse/californium/core/CoapObserveRelation.java

```diff
@@ -203,9 +203,8 @@ private void cancel() {
 	 * GET with Observe=1.
 	 */
 	public void proactiveCancel() {
-		sendCancelObserve();
-		// cancel observe relation
 		cancel();
+		sendCancelObserve();
```

Comment:
The other order was chosen, to ensure, that the token is not used in the "meantime" (between line 206 and 207) by an other request. But it's academically and not really analysed, if this is really a risk.

Reply:
I feel you're right that we may have a potential token collision here, but I think that's was already the case before because `sendCancelObserve` is not synchronous.

If we want to avoid this, the observationStore should contains the observation (and so lock the token) until we complete the proactive cancel exchange.

If we want to do that we should go with something more like you do in #694.

So, I will not integrate this last commit and merge only the tests.